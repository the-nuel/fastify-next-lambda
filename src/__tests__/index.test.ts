import { handler } from '../index';

describe('handler', () => {
    it('returns static content for static pages', async () => {
        const response = await handler({
            httpMethod: 'GET',
            path: '/',
            headers: {
                'x-nuel-traceid': 'static.content',
            },
        });

        expect(response).toMatchSnapshot({
            headers: {
                date: expect.any(String),
            },
        });
    });

    it('returns dynamic content for dynamic pages', async () => {
        const response = await handler({
            httpMethod: 'GET',
            path: '/about',
            headers: {
                'x-nuel-traceid': 'dynamic.content',
            },
        });

        expect(response).toMatchSnapshot({
            headers: {
                date: expect.any(String),
            },
        });
    });

    it('support dynamic routing', async () => {
        const response = await handler({
            httpMethod: 'GET',
            path: '/params/28/tom',
            headers: {
                'x-nuel-traceid': 'dynamic.routing',
            },
        });

        expect(response).toMatchSnapshot({
            headers: {
                date: expect.any(String),
            },
        });
    });

    it('returns the 404 page if there is no route match', async () => {
        const response = await handler({
            httpMethod: 'GET',
            path: '/some-nonexistant-path',
            headers: {
                'x-nuel-traceid': 'not.found',
            },
        });

        expect(response).toMatchSnapshot({
            headers: {
                date: expect.any(String),
            },
        });
    });
});

describe('caching', () => {
    beforeEach(() => {
        jest.resetModules();
    });

    afterEach(() => {
        delete process.env.STATIC_DEFAULT_MAX_AGE;
        delete process.env.DYNAMIC_DEFAULT_MAX_AGE;
    });

    it('defaults cache-control headers for static files to 0s', async () => {
        const handler = require('../index').handler;
        const response = await handler({ httpMethod: 'GET', path: '/' });
        expect(response.headers['cache-control']).toBe('public, max-age=0');
    });

    it('allows cache-control header duration for static files to be set', async () => {
        process.env.STATIC_DEFAULT_MAX_AGE = '60';
        const handler = require('../index').handler;
        const response = await handler({ httpMethod: 'GET', path: '/' });
        expect(response.headers['cache-control']).toBe('public, max-age=60');
    });

    it('defaults cache-control headers for dynamic files to 0s', async () => {
        const handler = require('../index').handler;
        const response = await handler({ httpMethod: 'GET', path: '/about' });
        expect(response.headers['cache-control']).toBe('public, max-age=0');
    });

    it('allows cache-control header duration for dynamic files to be set', async () => {
        process.env.DYNAMIC_DEFAULT_MAX_AGE = '30';
        const handler = require('../index').handler;
        const response = await handler({ httpMethod: 'GET', path: '/about' });
        expect(response.headers['cache-control']).toBe('public, max-age=30');
    });
});
