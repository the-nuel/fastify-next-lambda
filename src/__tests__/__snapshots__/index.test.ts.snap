// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler returns dynamic content for dynamic pages 1`] = `
Object {
  "body": "<html><head></head><body>This is placeholder output</body></html>",
  "headers": Object {
    "connection": "keep-alive",
    "content-security-policy": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "date": Any<String>,
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "transfer-encoding": "chunked",
    "x-content-security-policy": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-nuel-region": "unknown",
    "x-nuel-traceid": "dynamic.content",
    "x-webkit-csp": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "x-xss-protection": "1; mode=block",
  },
  "isBase64Encoded": false,
  "statusCode": 200,
}
`;

exports[`handler returns static content for static pages 1`] = `
Object {
  "body": "<!DOCTYPE html>
<html>
    <head>
        <title>This is a test page</title>
    </head>
    <body>
        This is some test content.
    </body>
</html>
",
  "headers": Object {
    "accept-ranges": "bytes",
    "cache-control": "public, max-age=0",
    "connection": "keep-alive",
    "content-length": 164,
    "content-security-policy": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "content-type": "text/html; charset=UTF-8",
    "date": Any<String>,
    "etag": "W/\\"a4-16c9be8f004\\"",
    "last-modified": "Fri, 16 Aug 2019 19:30:30 GMT",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "x-content-security-policy": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-nuel-region": "unknown",
    "x-nuel-traceid": "static.content",
    "x-webkit-csp": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "x-xss-protection": "1; mode=block",
  },
  "isBase64Encoded": false,
  "statusCode": 200,
}
`;

exports[`handler returns the 404 page if there is no route match 1`] = `
Object {
  "body": "<!DOCTYPE html>
<html>
    <head>
        <title>Not found</title>
    </head>
    <body>
        This page was not found.
    </body>
</html>
",
  "headers": Object {
    "accept-ranges": "bytes",
    "cache-control": "public, max-age=0",
    "connection": "keep-alive",
    "content-length": 152,
    "content-security-policy": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "content-type": "text/html; charset=UTF-8",
    "date": Any<String>,
    "etag": "W/\\"98-16c9bfed72a\\"",
    "last-modified": "Fri, 16 Aug 2019 19:54:25 GMT",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "x-content-security-policy": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-nuel-region": "unknown",
    "x-nuel-traceid": "not.found",
    "x-webkit-csp": "default-src 'self'; object-src 'none'; style-src 'self' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; upgrade-insecure-requests; block-all-mixed-content",
    "x-xss-protection": "1; mode=block",
  },
  "isBase64Encoded": false,
  "statusCode": 404,
}
`;
